FootprintIT tests fail with 'insights' profile

### Describe the bug

FootprintIT tests fail when Infinispan is built with `insights` profile due to inclusion of `infinispan-insights` client jar.  When the profile is activated or ideally when the project was built with the profile activated the test would take add an offset to expected results.

Here are size comparisons with and without the profile activated:
```
# FootPrintIT - no profile
-----------------------------
========== DISK Footprint: 75954300
========== CLASS Footprint: 12001
========== MEMORY Footprint: 27123480

# FootPrintIT - release-mode=downstream
-----------------------------
========== DISK Footprint: 76019809
========== CLASS Footprint: 12083
========== MEMORY Footprint: 27130920
```

### Expected behavior

Test will pass

### Actual behavior

Tests fail.

`main:`
```
[ERROR] Failures: 
[ERROR]   FootprintIT.testDiskFootprint:78 [Disk footprint] 
Expecting actual:
  76019809L
to be between:
  [74000000L, 76000000L]
```

`15.0.x:`
```
[Loaded class count] 
Expecting actual:
  11465
to be between:
  [11200, 11400]

[Disk footprint] 
Expecting actual:
  70009897L
to be between:
  [69000000L, 70000000L]
```

### How to Reproduce?

```
mvn clean install -Drelease-mode=downstream -DskipTests -s maven-settings.xml
mvn verify -Dit.test=FootprintIT -pl server/tests
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Infinispan version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/.gitignore
+++ b/.gitignore
@@ -44,3 +44,4 @@ classes
 # Checkstyle in Eclipse
 .checkstyle
 .idea
+jacoco

--- a/pom.xml
+++ b/pom.xml
@@ -2754,6 +2754,8 @@
                   <configuration>
                      <systemPropertyVariables>
                         <coverage.enabled>true</coverage.enabled>
+                        <version.jacoco>${version.jacoco}</version.jacoco>
+                        <org.infinispan.test.server.extension.libs>org.jacoco:org.jacoco.agent:${version.jacoco}:runtime</org.infinispan.test.server.extension.libs>
                      </systemPropertyVariables>
                   </configuration>
                </plugin>

--- a/server/testdriver/core/src/main/java/org/infinispan/server/test/core/ContainerInfinispanServerDriver.java
+++ b/server/testdriver/core/src/main/java/org/infinispan/server/test/core/ContainerInfinispanServerDriver.java
@@ -5,11 +5,11 @@ import static org.infinispan.server.Server.DEFAULT_SERVER_CONFIG;
 import static org.infinispan.server.test.core.Containers.DOCKER_CLIENT;
 import static org.infinispan.server.test.core.Containers.getDockerBridgeAddress;
 import static org.infinispan.server.test.core.Containers.imageArchitecture;
+import static org.infinispan.server.test.core.TestSystemPropertyNames.COVERAGE_ENABLED;
 import static org.infinispan.server.test.core.TestSystemPropertyNames.INFINISPAN_TEST_SERVER_CONTAINER_ULIMIT;
 import static org.infinispan.server.test.core.TestSystemPropertyNames.INFINISPAN_TEST_SERVER_CONTAINER_VOLUME_REQUIRED;
 import static org.infinispan.server.test.core.TestSystemPropertyNames.INFINISPAN_TEST_SERVER_LOG_FILE;
 import static org.infinispan.server.test.core.TestSystemPropertyNames.JACOCO_REPORTS_DIR;
-import static org.infinispan.server.test.core.TestSystemPropertyNames.COVERAGE_ENABLED;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
@@ -87,7 +87,7 @@ public class ContainerInfinispanServerDriver extends AbstractInfinispanServerDri
    private static final Long IMAGE_MEMORY = Long.getLong(TestSystemPropertyNames.INFINISPAN_TEST_SERVER_CONTAINER_MEMORY, null);
    private static final Long IMAGE_MEMORY_SWAP = Long.getLong(TestSystemPropertyNames.INFINISPAN_TEST_SERVER_CONTAINER_MEMORY_SWAP, null);
    public static final String INFINISPAN_SERVER_HOME = "/opt/infinispan";
-   public static final String JACOCO_COVERAGE_CONTAINER_PATH = "/opt/jacoco.exec";
+   public static final String JACOCO_COVERAGE_CONTAINER_PATH = INFINISPAN_SERVER_HOME + "/server/lib/jacoco.exec";
    public static final String JDK_BASE_IMAGE_NAME = "registry.access.redhat.com/ubi9/openjdk-21-runtime";
    private static final String[] IMAGE_DEPENDENCIES = {
          "file",
@@ -325,17 +325,17 @@ public class ContainerInfinispanServerDriver extends AbstractInfinispanServerDri
                   assert softHard.length == 2 : "Ulimit property must have format '<soft>,<hard>'";
                   long soft = Long.parseLong(softHard[0]);
                   long hard = Long.parseLong(softHard[1]);
-                  cmd.getHostConfig().withUlimits(new Ulimit[] { new Ulimit("nofile", soft, hard) });
+                  cmd.getHostConfig().withUlimits(new Ulimit[]{new Ulimit("nofile", soft, hard)});
                }
             });
       if (configuration.numServers() == 1 && (OS.getCurrentOs().equals(OS.MAC_OS) || OS.getCurrentOs().equals(OS.WINDOWS))) {
          container.addExposedPorts(
-                 11222, // HTTP/Hot Rod
-                 7800  // JGroups TCP
+               11222, // HTTP/Hot Rod
+               7800  // JGroups TCP
          );
       }
       String debug = configuration.properties().getProperty(TestSystemPropertyNames.INFINISPAN_TEST_SERVER_CONTAINER_DEBUG);
-      String javaOpts = null;
+      StringBuilder javaOpts = new StringBuilder();
       String site = configuration.site();
       if (site != null) {
          if (!sites.contains(site)) {
@@ -343,24 +343,24 @@ public class ContainerInfinispanServerDriver extends AbstractInfinispanServerDri
          }
       }
       if (i == 0 && site == null) {
-         javaOpts = "-D" + JOIN_TIMEOUT + "=0";
+         javaOpts.append("-D").append(JOIN_TIMEOUT).append("=0");
       }
       if (debug != null && Integer.parseInt(debug) == i) {
-         javaOpts = javaOpts == null ? debugJvmOption() : javaOpts + " " + debugJvmOption();
+         javaOpts.append(" ").append(debugJvmOption());
          log.infof("Container debug enabled with options '%s'%n", javaOpts);
       }
       String isCoverageEnabled = System.getProperty(COVERAGE_ENABLED);
       if (Boolean.parseBoolean(isCoverageEnabled)) {
-         javaOpts = javaOpts == null ? "-javaagent:/opt/infinispan/server/lib/org.jacoco.agent-0.8.12-runtime.jar=output=file,destfile=" + JACOCO_COVERAGE_CONTAINER_PATH + ",append=true"
-                 : javaOpts + " " + "-javaagent:/opt/infinispan/server/lib/org.jacoco.agent-0.8.12-runtime.jar=output=file,destfile=" + JACOCO_COVERAGE_CONTAINER_PATH + ",append=true";
+         String jacocoVersion = System.getProperty("version.jacoco");
+         javaOpts.append(" -javaagent:/opt/infinispan/server/lib/org.jacoco.agent-").append(jacocoVersion).append("-runtime.jar=output=file,destfile=").append(JACOCO_COVERAGE_CONTAINER_PATH).append(",append=true");
       }
 
-      if (javaOpts != null) {
+      if (!javaOpts.isEmpty()) {
          String baseImage = configuration.properties().getProperty(TestSystemPropertyNames.INFINISPAN_TEST_SERVER_BASE_IMAGE_NAME);
          if (baseImage == null) {
-            container.withEnv("JAVA_OPTS", javaOpts);
+            container.withEnv("JAVA_OPTS", javaOpts.toString());
          } else {
-            container.withEnv("JAVA_OPTIONS", javaOpts);
+            container.withEnv("JAVA_OPTIONS", javaOpts.toString());
          }
       }
 
@@ -505,7 +505,7 @@ public class ContainerInfinispanServerDriver extends AbstractInfinispanServerDri
          RMIServer stub = (RMIServer) registry.lookup("jmxrmi");
          JMXConnector connector = new RMIConnector(stub, null);
          Map<String, Object> env = new HashMap<>();
-         env.put(JMXConnector.CREDENTIALS, new String[] {username, password});
+         env.put(JMXConnector.CREDENTIALS, new String[]{username, password});
          connector.connect(env);
          log.infof("Connecting to JMX URL %s", url.toString());
          reaper.accept(connector);
@@ -608,8 +608,13 @@ public class ContainerInfinispanServerDriver extends AbstractInfinispanServerDri
    static String createServerImage(String serverOutputDir) {
       try {
          InspectImageResponse response = DOCKER_CLIENT.inspectImageCmd(SNAPSHOT_IMAGE).exec();
-         log.infof("Reusing existing image");
-         return response.getConfig().getImage();
+         log.infof("Reusing existing image: %s [%s]", SNAPSHOT_IMAGE, response.getConfig());
+         String name = response.getConfig().getImage();
+         if (name == null || name.isEmpty()) {
+            return SNAPSHOT_IMAGE;
+         } else {
+            return name;
+         }
       } catch (NotFoundException e) {
          // We build our local image based on the supplied server directory
          Path serverOutputPath = Paths.get(serverOutputDir).normalize();

--- a/server/tests/pom.xml
+++ b/server/tests/pom.xml
@@ -283,6 +283,18 @@
                <scope>test</scope>
             </dependency>
          </dependencies>
+         <build>
+            <plugins>
+               <plugin>
+                  <artifactId>maven-failsafe-plugin</artifactId>
+                  <configuration>
+                     <systemPropertyVariables>
+                        <insights.enabled>true</insights.enabled>
+                     </systemPropertyVariables>
+                  </configuration>
+               </plugin>
+            </plugins>
+         </build>
       </profile>
    </profiles>
 </project>

--- a/server/tests/src/test/java/org/infinispan/server/footprint/FootprintIT.java
+++ b/server/tests/src/test/java/org/infinispan/server/footprint/FootprintIT.java
@@ -26,12 +26,20 @@ import org.junit.jupiter.api.extension.RegisterExtension;
  * @since 15.0
  **/
 public class FootprintIT {
-   private static final int LOADED_CLASS_COUNT_LOWER_BOUND = 11_700;
-   private static final int LOADED_CLASS_COUNT_UPPER_BOUND = 11_900;
-   private static final long HEAP_USAGE_LOWER_BOUND = 25_000_000L;
-   private static final long HEAP_USAGE_UPPER_BOUND = 28_000_000L;
-   private static final long DISK_USAGE_LOWER_BOUND = 74_000_000L;
+   private static final int LOADED_CLASS_COUNT_LOWER_BOUND = 11_500;
+   private static final int LOADED_CLASS_COUNT_UPPER_BOUND = 12_000;
+   private static final long HEAP_USAGE_LOWER_BOUND = 26_000_000L;
+   private static final long HEAP_USAGE_UPPER_BOUND = 27_000_000L;
+   private static final long DISK_USAGE_LOWER_BOUND = 75_000_000L;
    private static final long DISK_USAGE_UPPER_BOUND = 76_000_000L;
+
+   private static final int JACOCO_CLASS_COUNT = 165;
+   private static final long JACOCO_HEAP_USAGE = 1_900_000;
+
+   private static final int INSIGHTS_CLASS_COUNT = 90;
+   private static final int INSIGHTS_HEAP_USAGE = 55_000;
+   private static final long INSIGHTS_DISK_USAGE = 250_000;
+
    public static final String HEAP_DUMP = "footprint.hprof";
 
    @RegisterExtension
@@ -49,10 +57,20 @@ public class FootprintIT {
       ObjectName memory = new ObjectName("java.lang:type=Memory");
       jmxConnection.invoke(memory, "gc", new Object[0], new String[0]);
       CompositeData heapMemoryUsage = (CompositeData) jmxConnection.getAttribute(memory, "HeapMemoryUsage");
-      Long used = (Long) heapMemoryUsage.get("used");
+      Long usedHeap = (Long) heapMemoryUsage.get("used");
+      int classCountOffset = 0;
+      long heapCountOffset = 0;
+      if (Boolean.getBoolean("coverage.enabled")) {
+         classCountOffset += JACOCO_CLASS_COUNT;
+         heapCountOffset += JACOCO_HEAP_USAGE;
+      }
+      if (Boolean.getBoolean("insights.enabled")) {
+         classCountOffset += INSIGHTS_CLASS_COUNT;
+         heapCountOffset += INSIGHTS_HEAP_USAGE;
+      }
       try {
-         assertThat(loadedClassCount).as("Loaded class count").isBetween(LOADED_CLASS_COUNT_LOWER_BOUND, LOADED_CLASS_COUNT_UPPER_BOUND);
-         assertThat(used).as("Heap memory usage").isBetween(HEAP_USAGE_LOWER_BOUND, HEAP_USAGE_UPPER_BOUND);
+         assertThat(loadedClassCount - classCountOffset).as("Loaded class count").isBetween(LOADED_CLASS_COUNT_LOWER_BOUND, LOADED_CLASS_COUNT_UPPER_BOUND);
+         assertThat(usedHeap - heapCountOffset).as("Heap memory usage").isBetween(HEAP_USAGE_LOWER_BOUND, HEAP_USAGE_UPPER_BOUND);
       } catch (AssertionError e) {
          ObjectName hotSpot = new ObjectName("com.sun.management:type=HotSpotDiagnostic");
          jmxConnection.invoke(hotSpot, "dumpHeap", new Object[]{HEAP_DUMP, true}, new String[]{"java.lang.String", "boolean"});
@@ -64,12 +82,16 @@ public class FootprintIT {
 
    @Test
    public void testDiskFootprint() throws IOException {
+      long diskCountOffset = 0;
+      if (Boolean.getBoolean("insights.enabled")) {
+         diskCountOffset += INSIGHTS_DISK_USAGE;
+      }
       Path folder = Paths.get(System.getProperty("org.infinispan.test.server.dir"));
       try (Stream<Path> stream = Files.walk(folder)) {
          long size = stream.filter(p -> p.toFile().isFile())
                .mapToLong(p -> p.toFile().length())
                .sum();
-         assertThat(size).as("Disk footprint").isBetween(DISK_USAGE_LOWER_BOUND, DISK_USAGE_UPPER_BOUND);
+         assertThat(size - diskCountOffset).as("Disk footprint").isBetween(DISK_USAGE_LOWER_BOUND, DISK_USAGE_UPPER_BOUND);
       }
    }
 }

